Rover/Map.cpp

```diff
@@ -0,0 +1,327 @@
+    float R = sqrt(diff.first * diff.first + diff.second * diff.second);
+    //</hack>
+    
+    auto direction = std::make_pair(end.first - start.first,
+				    end.second - start.second);
+    auto center_to_start = std::make_pair(start.first - circle.first,
+					  start.second - circle.second);
+    float a = direction.first * direction.first + direction.second * direction.second;
+    float b = 2 * (center_to_start.first * direction.first + center_to_start.second * direction.second);
+    float c = (center_to_start.first * center_to_start.first + center_to_start.second * center_to_start.second) + R * R;
+
+    float descriminant = b * b - 4 * a * c;
+    if(descriminant < 0)
+    {
+	return(false);
+    }
+
+    descriminant = sqrt(descriminant);
+    float t1 = (-b + descriminant) / (2 * a);
+    float t2 = (-b - descriminant) / (2 * a);
+
+    return((0 <= t1 && t1 <= 1.0f) || (0 <= t2 && t2 <= 1.0f));	
+}
+
+//Returns 0 if p, q, and r are colinear.
+//Returns 1 if pq, qr, and rp are clockwise
+//Returns 2 if pq, qr, and rp are counterclockwise
+int RoverPathfinding::Map::orientation(point p, point q, point r)
+{
+    float v = (q.second - p.second) * (r.first - q.first) -
+	(q.first - p.first) * (r.second - q.second);
+    if(-1e-7 <= v && v <= 1e-7)
+	return 0;
+
+    return((v > 0.0f) ? 1 : 2);
+}
+
+bool RoverPathfinding::Map::on_segment(point p, point q, point r)
+{
+    return(q.first <= std::max(p.first, r.first) && q.first >= std::min(p.first, r.first) &&
+	   q.second <= std::max(p.second, r.second) && q.second >= std::min(p.second, r.second));
+}
+
+bool RoverPathfinding::Map::segments_intersect(point p1, point p2, point q1, point q2)
+{
+    int o1 = orientation(p1, q1, p2);
+    int o2 = orientation(p1, q1, q2);
+    int o3 = orientation(p2, q2, p1);
+    int o4 = orientation(p2, q2, q1);
+ 
+    if (o1 != o2 && o3 != o4)
+        return(true);
+ 
+    if (o1 == 0 && on_segment(p1, p2, q1))
+	return(true);
+    
+    if (o2 == 0 && on_segment(p1, q2, q1))
+	return(true);
+ 
+    if (o3 == 0 && on_segment(p2, p1, q2))
+	return(true);
+ 
+    if (o4 == 0 && on_segment(p2, q1, q2))
+	return(true);
+ 
+    return(false);
+}
+
+std::pair<RoverPathfinding::point, RoverPathfinding::point> RoverPathfinding::Map::add_length_to_line_segment(point p, point q, float length)
+{
+    point pq = std::make_pair(q.first - p.first, q.second - p.second); //vector
+    float len = sqrt(pq.first * pq.first + pq.second * pq.second);
+    pq.first = length * pq.first / len;
+    pq.second = length * pq.second / len;
+
+    point p1 = std::make_pair(q.first + pq.first, q.second + pq.second);
+    point p2 = std::make_pair(p.first - pq.first, p.second - pq.second);
+    return(std::make_pair(p1, p2));   
+}
+
+RoverPathfinding::point RoverPathfinding::Map::center_point_with_radius(RoverPathfinding::point cur, RoverPathfinding::point p, RoverPathfinding::point q, float R)
```

Comment:
I'm having trouble figuring out what this does, maybe include a comment, or change the name?